docs/bounded_dynamism.md

```diff
@@ -0,0 +1,85 @@
+# StableHLO Bounded Dynamism
+
+[StableHLO](https://github.com/openxla/stablehlo) is an operation set that expresses ML computations. It has been originally bootstrapped from [the MHLO dialect](https://github.com/tensorflow/mlir-hlo#meta-hlo-dialect-mhlo), including inheriting the type and some of the ops. This RFC aims to describe the current status and future plans for bounded dynamism in StableHLO. In particular, this RFC doesn’t propose any further major changes to the current state. The only minor change proposed is allowing `i64` typed result for `get_dimension_size` op and `i64` type for `set_dimension_size` `size` operand. This is further described in P3.
+
+Bounded dynamism allows computations to encode the maximum runtime size of dynamic dimensions of tensors. This makes it possible to run particular computations on hardware that doesn’t support dynamic tensors and needs hints on the largest possible. For example, real time inference might require support for dynamic batch so that inference doesn’t need to wait for accumulation up to the static batch size. Similarly, during training the last batch in a dataset might be smaller than the max batch size if the number of examples in the dataset is not divisible by the batch size.
+
+
+# Proposal
+
+
+## (P1) Use RankedTensorType encoding field for bounds
+
+Bounds for a dynamic tensor are represented using the `TypeExtensionsAttr` using the encoding field in the `RankedTensorType`. Bounds in `TypeExtensionsAttr` is an `int64_t` array of size equal to rank of the tensor. Values corresponding to static dimensions must be `ShapedType::kDynamicSize` which is `-1`.
```

Comment:
Use of `-1` strictly is likely to lead to implementation issues in the future. In pytorch side, we landed on a more relaxed scheme to arrive at the same behavior. https://github.com/pytorch/pytorch/blob/master/c10/core/SymInt.h#L197-L218

Operations like `torch.expand` have explicit use of `-1` that prevent us from relying on `-1`. Let's revisit if we can adjust this requirement. Wdyt?

Reply:
Good point! `-1` was used only while printing the IR. Value is ShapedType::kDynamicSize. Now, it is printed as `?`.

Reply:
I see the following comment in the example provided in the [RFC](https://github.com/openxla/stablehlo/blob/71d01753bf79a126eb1a8d5ead03d456689b8d8e/rfcs/20220925-bounded-dynamism.md):

> `  // Conditionally add size of the input data if the indices are negative.`

Is there a special behavior reserved for for negative indices still?

Reply:
seems like you are looking at an older version. Newer version shouldn't have it.